src/batRandom.ml

```diff
@@ -73,6 +73,37 @@ let enum_char () = BatEnum.from char
 
 let choice e = BatEnum.drop (int (BatEnum.count e)) e; BatEnum.get_exn e
 
+let weighted_choice e =
+  let list, total_weight =
+    BatEnum.fold (fun (acc, total) ((_, w) as elem) ->
+        if w < 0. then
+          Printf.kprintf invalid_arg
+            "weighted_choice does not accept negative weights (here %f)"
+            w;
+        elem :: acc, total +. w) ([], 0.) e in
+
+  if list = [] || total_weight <= 0. then raise Not_found;
+
+  let pick = float total_weight in
+  let rec find level = function
+    | [] -> assert false
```

Comment:
This case is in fact statically known to be impossible, hence the `assert false`. I added a comment to justify this impossibility. It is always good practice to do so, thanks for the remark.